### PR TITLE
feat: add bft function testcases in new substrait testfile format

### PR DESCRIPTION
**Pull request by srikrishnak**

### CODE REVIEW

**Comment by jacques-n** [comment]
We should get CI running first for the coverage code before adding these. That way we won't accidentally add broken things.

Second, we should also add the automated tool that does the conversion so other people can use it if they have their own tests.

File: tests/cases/datetime/add_datetime.test
Comment (EpsilonPrime) [inline tests/cases/datetime/add_datetime.test:1]: add_intervals is missing

File: tests/cases/datetime/extract.test [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+# date: examples using the date type
+extract('YEAR'::str, '2020-12-31'::date) = 2020::i64
+extract('MONTH'::str, '2020-12-31'::date) = 12::i64
+extract('DAY'::str, '2020-12-31'::date) = 31::i64
+
+# time: examples using the time type
+extract('HOUR'::str, '01:02:03.155900'::time) = 1::i64
+extract('MINUTE'::str, '01:02:03.155900'::time) = 2::i64
+extract('SECOND'::str, '01:02:03.155900'::time) = 3::i64
+extract('MILLISECOND'::str, '01:02:03.155900'::time) = 3000::i64
```
Comment (EpsilonPrime) [inline tests/cases/datetime/extract.test:33]: The format in the original was 01:02:03+1559.  I don't believe the correct answer for 01:02:03.155900 is 3000.  If this only includes the fractional part the answer ought to be 155 or 156.

File: tests/cases/datetime/extract.test [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+extract('MICROSECOND'::str, '01:02:03.45'::time) = 3450000::i64
```
Comment (EpsilonPrime) [inline tests/cases/datetime/extract.test:34]: MySQL returns only the fractional part so I wouldn't expect the correct value to include 3 seconds.

File: tests/cases/datetime/gt_datetime.test [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+gt('2018-12-31T13:30:15'::ts, '2017-12-31T13:30:15'::ts) = true::bool
+
+# timestamp_tz: examples using the timestamp_tz type
+gt('1999-01-08T01:05:05-08:00'::tstz, '1999-01-08T04:05:06-05:00'::tstz) = false::bool
+gt('1999-01-08T01:05:07-08:00'::tstz, '1999-01-08T04:05:06-05:00'::tstz) = true::bool
+
+# date: examples using the date type
+gt('2020-12-30'::date, '2020-12-31'::date) = false::bool
+gt('2020-12-31'::date, '2020-12-30'::date) = true::bool
+gt('P7D'::iday, 'P6D'::iday) = true::bool
```
Comment (EpsilonPrime) [inline tests/cases/datetime/gt_datetime.test:15]: move to next group

File: tests/cases/datetime/gte_datetime.test [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+# timestamp_tz: examples using the timestamp_tz type
+gte('1999-01-08T01:05:05-08:00'::tstz, '1999-01-08T04:05:06-05:00'::tstz) = false::bool
+gte('1999-01-08T01:05:06-08:00'::tstz, '1999-01-08T01:05:06-08:00'::tstz) = true::bool
+gte('1999-01-08T01:05:06-08:00'::tstz, '1999-01-08T04:05:05-05:00'::tstz) = true::bool
+
+# date: examples using the date type
+gte('2020-12-30'::date, '2020-12-31'::date) = false::bool
+gte('2020-12-31'::date, '2020-12-31'::date) = true::bool
+gte('2020-12-31'::date, '2020-12-30'::date) = true::bool
+gte('P7D'::iday, 'P7D'::iday) = true::bool
```
Comment (EpsilonPrime) [inline tests/cases/datetime/gte_datetime.test:18]: move below

File: tests/cases/datetime/lt_datetime.test [hunk 5]
```diff
@@ -12,9 +12,9 @@ lt('1999-01-08T01:05:06-08:00'::tstz, '1999-01-08T04:05:06-05:00'::tstz) = false
 # date: examples using the date type
 lt('2020-12-30'::date, '2020-12-31'::date) = true::bool
 lt('2020-12-31'::date, '2020-12-30'::date) = false::bool
+lt('P7D'::iday, 'P6D'::iday) = false::bool
```
Comment (EpsilonPrime) [inline tests/cases/datetime/lt_datetime.test:15]: move below

File: tests/cases/datetime/lte_datetime.test [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+# timestamp_tz: examples using the timestamp_tz type
+lte('1999-01-08T01:05:05-08:00'::tstz, '1999-01-08T04:05:06-05:00'::tstz) = true::bool
+lte('1999-01-08T01:05:06-08:00'::tstz, '1999-01-08T01:05:06-08:00'::tstz) = true::bool
+lte('1999-01-08T01:05:06-08:00'::tstz, '1999-01-08T04:05:05-05:00'::tstz) = false::bool
+
+# date: examples using the date type
+lte('2020-12-30'::date, '2020-12-31'::date) = true::bool
+lte('2020-12-31'::date, '2020-12-31'::date) = true::bool
+lte('2020-12-31'::date, '2020-12-30'::date) = false::bool
+lte('P7D'::iday, 'P7D'::iday) = true::bool
```
Comment (EpsilonPrime) [inline tests/cases/datetime/lte_datetime.test:18]: move below

File: tests/cases/logarithmic/ln.test [hunk 7]
```diff
@@ -0,0 +1,19 @@
+### SUBSTRAIT_SCALAR_TEST: v1.0
+### SUBSTRAIT_INCLUDE: '/extensions/functions_logarithmic.yaml'
+
+# basic: Basic examples without any special cases
+ln(100000::i64) = 11.512925464970229::fp64
+ln(1.0::fp32) = 0::fp32
+ln(2.015::fp64) = 0.7006191953986464::fp64
+ln(inf::fp64) = inf::fp64
```
Comment (EpsilonPrime) [inline tests/cases/logarithmic/ln.test:8]: duplicate of an infinity test

File: tests/cases/logarithmic/log10.test [hunk 8]
```diff
@@ -0,0 +1,19 @@
+### SUBSTRAIT_SCALAR_TEST: v1.0
+### SUBSTRAIT_INCLUDE: '/extensions/functions_logarithmic.yaml'
+
+# basic: Basic examples without any special cases
+log10(100000::i64) = 5.0::fp64
+log10(1.0::fp32) = 0::fp32
+log10(2.015::fp64) = 0.3042750504771283::fp64
+log10(inf::fp64) = inf::fp64
```
Comment (EpsilonPrime) [inline tests/cases/logarithmic/log10.test:8]: duplicate of an infinity test

File: tests/cases/logarithmic/log2.test [hunk 9]
```diff
@@ -0,0 +1,20 @@
+### SUBSTRAIT_SCALAR_TEST: v1.0
+### SUBSTRAIT_INCLUDE: '/extensions/functions_logarithmic.yaml'
+
+# basic: Basic examples without any special cases
+log2(100000::i64) = 16.609640474436812::fp64
+log2(1.0::fp32) = 0::fp32
+log2(8.0::fp64) = 3.0::fp64
+log2(2.015::fp64) = 1.0107798387532427::fp64
+log2(inf::fp64) = inf::fp64
```
Comment (EpsilonPrime) [inline tests/cases/logarithmic/log2.test:9]: duplicate of an infinity test

File: tests/cases/string/bit_length.test [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+bit_length(' '::str) = 8::i64
+bit_length('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'::str) = 384::i64
+bit_length('   456'::str) = 48::i64
+
+# null_input: Examples with null as input
+bit_length(null::str) = null::i64
+
+# unicode: Examples with unicode characters as input
+bit_length('à'::str) = 16::i64
+bit_length('😄'::str) = 32::i64
```
Comment (EpsilonPrime) [inline tests/cases/string/bit_length.test:16]: future:  might be useful to add a three byte unicode character test to all the length functions

File: tests/cases/string/contains.test [hunk 11]
```diff
@@ -0,0 +1,14 @@
+### SUBSTRAIT_SCALAR_TEST: v1.0
+### SUBSTRAIT_INCLUDE: '/extensions/functions_string.yaml'
+
+# basic: Basic examples contains as prefix
+contains('abcdefg'::str, 'abc'::str) = true::bool
+contains('abcdefg'::str, 'CdE'::str) [case_sensitivity:CASE_INSENSITIVE] = true::bool
```
Comment (EpsilonPrime) [inline tests/cases/string/contains.test:6]: would also be useful to test what happens when case insensitivity is not specified for the same input

File: tests/cases/string/ends_with.test [hunk 12]
```diff
@@ -0,0 +1,15 @@
+### SUBSTRAIT_SCALAR_TEST: v1.0
+### SUBSTRAIT_INCLUDE: '/extensions/functions_string.yaml'
+
+# basic: Basic examples without any special cases
+ends_with('abcd'::str, 'd'::str) = true::bool
+ends_with('abcd'::str, 'a'::str) = false::bool
+
+# case_insenstivity: multi byte character comparison with case insensitivity
+ends_with('abcd'::str, 'CD'::str) [case_sensitivity:CASE_INSENSITIVE] = true::bool
```
Comment (EpsilonPrime) [inline tests/cases/string/ends_with.test:9]: add a test for the same input but without case insensitivity

File: tests/cases/string/like.test [hunk 13]
```diff
@@ -0,0 +1,12 @@
+### SUBSTRAIT_SCALAR_TEST: v1.0
+### SUBSTRAIT_INCLUDE: '/extensions/functions_string.yaml'
+
+# basic: Basic examples without any special cases
+like('abcdefg'::str, 'abcdefg'::str) = true::bool
+like('abcdefg'::str, 'abc'::str) = false::bool
+
```
Comment (EpsilonPrime) [inline tests/cases/string/like.test:7]: future: add a test case with uppercase to verify that wildcards are case sensitive

File: tests/cases/string/lower.test [hunk 14]
```diff
@@ -0,0 +1,15 @@
+### SUBSTRAIT_SCALAR_TEST: v1.0
+### SUBSTRAIT_INCLUDE: '/extensions/functions_string.yaml'
+
+# basic: Basic examples without any special cases
+lower('ABC'::str) = 'abc'::str
+lower('aBc'::str) = 'abc'::str
+lower('abc'::str) = 'abc'::str
+lower(''::str) = ''::str
+
```
Comment (EpsilonPrime) [inline tests/cases/string/lower.test:9]: future:  we are going to need to revisit all of the string cases once we add collation and character set.  (For instance, the capital letter I in Turkish does not become i but a version without a dot.)

File: tests/cases/string/regexp_replace.test
Comment (EpsilonPrime) [inline tests/cases/string/regexp_replace.test:1]: regexp_string_split is missing

File: tests/cases/string/repeat.test [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+
+# basic: Basic examples without any special cases
+repeat('abc'::str, 2::i64) = 'abcabc'::str
+repeat('aBc'::str, 0::i64) = ''::str
+repeat('   abd'::str, 3::i64) = '   abd   abd   abd'::str
+repeat(' '::str, 5::i64) = '     '::str
+repeat(''::str, 2::i64) = ''::str
+
+# null_input: Examples with null as input
+repeat(null::str, 2::i64) = null::str
```
Comment (EpsilonPrime) [inline tests/cases/string/repeat.test:12]: later:  add tests for null and negative counts

File: tests/cases/string/starts_with.test [hunk 16]
```diff
@@ -0,0 +1,15 @@
+### SUBSTRAIT_SCALAR_TEST: v1.0
+### SUBSTRAIT_INCLUDE: '/extensions/functions_string.yaml'
+
+# basic: Basic examples without any special cases
+starts_with('abcd'::str, 'a'::str) = true::bool
+starts_with('abcd'::str, 'z'::str) = false::bool
+
+# case_insenstivity: multi byte character comparison with case insensitivity
+starts_with('abcd'::str, 'AB'::str) [case_sensitivity:CASE_INSENSITIVE] = true::bool
```
Comment (EpsilonPrime) [inline tests/cases/string/starts_with.test:9]: also needs a test with the same input but not case insensitive

**Review by EpsilonPrime — COMMENTED** [review]
This turned into a review of the BFT source testcases since the conversion process is working correctly.

**Comment by srikrishnak** [comment]
I took care of review comments but had to rebase because of conflicting file.
After rebase, I new counter in coverage pointed to around 400 signature matches although the testcases would pass on bft.

Below are the different cases which caused signature mismatches even for functions which have the right test cases, addressed almost all of them (4 is not yet fixed). Now I am left with around 250 odd test cases. Now, mostly I see failures which are actual signature mismatches or (4.)

1. tests with return type SubstraitError()
2. tests for functions like or having variadic arguments
3. tests with return type  decimal
4. tests with return type any1
5. tests with return type any

**Comment by srikrishnak** [comment]
> I took care of review comments but had to rebase because of conflicting file. After rebase, I new counter in coverage pointed to around 400 signature matches although the testcases would pass on bft.
> 
> Below are the different cases which caused signature mismatches even for functions which have the right test cases, addressed almost all of them (4 is not yet fixed). Now I am left with around 250 odd test cases. Now, mostly I see failures which are actual signature mismatches or (4.)
> 
> 1. tests with return type SubstraitError()
> 2. tests for functions like or having variadic arguments
> 3. tests with return type  decimal
> 4. tests with return type any1
> 5. tests with return type any

raised PR https://github.com/substrait-io/substrait/pull/744 to fix issues with function lookup in coverage tool. 
This PR will need rebase on https://github.com/substrait-io/substrait/pull/744

File: tests/cases/aggregate_generic/count.test
[the remark this answers is not in the document]
Comment (srikrishnak) [reply tests/cases/aggregate_generic/count.test:1]: The count_star() function name doesn't match any function in substrait functions.  So, unable to add it as coverage breaks.

**Review by EpsilonPrime — APPROVED** [review]
There's still the question of extract with microsecond but since that's mirroring the existing test it's fine to investigate it later.

File: tests/cases/datetime/extract.test (hunk 2 above)
Comment (srikrishnak) [reply tests/cases/datetime/extract.test:31]: The test_sqllite in bft skips extract as unsupported function.

**Comment by srikrishnak** [comment]
> I raised a draft PR [substrait-io/bft#98](https://github.com/substrait-io/bft/pull/98) to check how pipeline runs go with these testcases on bft.

I checked locally by using the DRAFT PR for duckdb, snowflake, postgres and sqllite. Made sure there are no regressions on them. 

The previous patch had one offending test case for sqllite, removed that testcase.

**Review by EpsilonPrime — APPROVED** [review]

**Review by jacques-n — APPROVED** [review]
Great work @srikrishnak and @EpsilonPrime ! Thanks for thorough update and review.
